### PR TITLE
MAINT: Bump ruff, fix E721 linting errors

### PR DESCRIPTION
**Pull request by connortann**
Update the ruff linter, and fix a few linting issues due to comparison of types. Also add a minor rule from flake8-simplify.

### CODE REVIEW

File: shap/explainers/_deep/deep_pytorch.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def __init__(self, model, data):
         self.interim = False
         self.interim_inputs_shape = None
         self.expected_value = None  # to keep the DeepExplainer base happy
-        if type(model) == tuple:
+        if isinstance(model, tuple):
```
Comment (CloseChoice) [inline shap/explainers/_deep/deep_pytorch.py:29]: this is an E721 fix? If so I am missing this error code in E721 to be ignored previously or why is this fix in here?

Comment (connortann) [reply shap/explainers/_deep/deep_pytorch.py:29]: I believe the behaviour of this rule has changed slightly in ruff 5.0:

> [type-comparison](https://docs.astral.sh/ruff/rules/type-comparison/) (E721) now allows idioms such as type(x) is int

**Comment by connortann** [comment]
On the topic of ruff rules I was looking into flake8-simplify, which has a few potentially useful rules. I added one rule to simplify `isinstance` calls. I was initially considering getting further rules passing, but they seemed less helpful / more pedantic than I originally hoped so I'll just leave it at the one.

**Review by CloseChoice — APPROVED** [review]
LGTM

**Comment by CloseChoice** [comment]
> On the topic of ruff rules I was looking into flake8-simplify, which has a few potentially useful rules. I added one rule to simplify `isinstance` calls. I was initially considering getting further rules passing, but they seemed less helpful / more pedantic than I originally hoped so I'll just leave it at the one.

Sounds reasonable. I like our current approach: taking the useful side of things but not being pedantic about these things.

**Pull request merged**